Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -519,6 +524,117 @@ extension SwiftLanguageServer {
+        req.reply(.failure(result.failure!))
+        return
+      }
+
+      guard let syntaxMap: SKResponseArray = dict[self.keys.syntaxmap],
+            let substructure: SKResponseArray = dict[self.keys.substructure] else {
+        return req.reply([])
+      }
+
+      var ranges: [FoldingRange] = []
+
+      // Merge successive comments into one big comment by adding their lengths.
+      var currentComment: (offset: Int, length: Int)? = nil
+
+      syntaxMap.forEach { _, value in
+        if let kind: sourcekitd_uid_t = value[self.keys.kind],
+           kind.isCommentKind(self.values),
+           let offset: Int = value[self.keys.offset],
+           let length: Int = value[self.keys.length]
+        {
+          if let comment = currentComment {
+            if offset == comment.offset + comment.length {
+              currentComment?.length += length
+            } else {
+              self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toArray: &ranges)
+              currentComment = (offset: offset, length: length)
+            }
+          } else {
+            currentComment = (offset: offset, length: length)
+          }
+        }
+        return true
+      }
+
+      // Add the last stored comment.
+      if let comment = currentComment {
+        self.addFoldingRange(offset: comment.offset, length: comment.length, kind: .comment, in: snapshot, toArray: &ranges)
+        currentComment = nil
+      }
+
+      var structureStack: [SKResponseArray] = [substructure]
+      while let substructure = structureStack.popLast() {
+        substructure.forEach { _, value in
+          if let offset: Int = value[self.keys.bodyoffset],
+             let length: Int = value[self.keys.bodylength],
+             length > 0
+          {
+            self.addFoldingRange(offset: offset, length: length, in: snapshot, toArray: &ranges)
+          }
+          if let substructure: SKResponseArray = value[self.keys.substructure] {
+            structureStack.append(substructure)
+          }
+          return true
+        }
+      }
+
+      req.reply(ranges)
+    }
+
+    // FIXME: cancellation
+    _ = handle
+  }
+
+  func addFoldingRange(offset: Int, length: Int, kind: FoldingRangeKind? = nil, in snapshot: DocumentSnapshot, toArray ranges: inout [FoldingRange]) {
+    let capabilities = clientCapabilities.textDocument?.foldingRange
+    if let rangeLimit = capabilities?.rangeLimit, ranges.count == rangeLimit {
+      return
+    }
+    guard let start: Position = snapshot.positionOf(utf8Offset: offset),
+          let end: Position = snapshot.positionOf(utf8Offset: offset + length) else {
+      return
+    }
+    let range: FoldingRange
+    // If the client only supports folding full lines, ignore the end character's line.
+    if capabilities?.lineFoldingOnly == true {
+      let lastLineToFold = end.line - 1
+      if lastLineToFold <= start.line {
```

Comment:
So you can never fold a single line?  Is that expected?

Reply:
If the client only supports folding entire lines but the range starts and ends at the same line, then you'll certainly end up folding things outside the range, such as the braces themselves. 

VSCode itself ignores the lines that start/end a range as it folds (startLine+1...endLine), with startLine being used only as a place to display the fold button and endLine being the line before the closing brace, so a range that starts and ends at the same line has no effect.